bigtable/client/detail/readrowsparser.h

```diff
@@ -0,0 +1,142 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_DETAIL_READROWSPARSER_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_DETAIL_READROWSPARSER_H_
+
+#include "bigtable/client/version.h"
+
+#include "bigtable/client/cell.h"
+#include "bigtable/client/row.h"
+
+#include <google/bigtable/v2/bigtable.grpc.pb.h>
+
+#include <absl/strings/string_view.h>
+#include <vector>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Transforms a stream of chunks as returned by the ReadRows streaming
+ * RPC into a sequence of rows.
+ *
+ * A simplified example of correctly using this class:
+ *
+ * @code
+ * while (!stream.EOT()) {
+ *   chunk = stream.NextChunk();
+ *   parser.HandleChunk(chunk);
+ *   if (parser.HasNext()) {
+ *     row = parser.Next();  // you now own `row`
+ *   }
+ * }
+ * parser.HandleEOT();
+ * @endcode
+ *
+ * NO RECYCLING of the parser object: This is a stateful class, and a
+ * single and unique parser should be used for each stream of ReadRows
+ * responses. If errors occur, an exception is thrown as documented by
+ * each method and the parser object is left in an undefined state.
+ */
+class ReadRowsParser {
+ public:
+  ReadRowsParser()
+      : row_key_(""),
+        cells_(),
+        cell_first_chunk_(true),
+        cell_(),
+        last_seen_row_key_(""),
+        row_ready_(false),
+        eot_(false) {}
+
+  /**
+   * Pass an input chunk proto to the parser.
+   *
+   * @throws std::runtime_error if called while a row is available
+   * (HasNext() is true).
+   *
+   * @throws std::runtime_error if validation failed.
+   */
+  void HandleChunk(google::bigtable::v2::ReadRowsResponse_CellChunk chunk);
+
+  /**
+   * Signal that the input stream reached the end.
+   *
+   * @throws std::runtime_error if more data was expected to finish
+   * the current row.
+   */
+  void HandleEOT();
```

Comment:
done
It was inspired by the EOT character (end of transmission), but since we are handling a streaming RPC I replaced it throughout with "End of stream".